### ISSUE TITLE
Infrequent Pragmatic Lambda Blog - Type Enthusiast's Notes about TypeScript. Part 4. Programming with Type Variables

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8793]
Infrequent, Pragmatic, Lambda Blog - Type Enthusiast's Notes about TypeScript. Part 4. Programming with Type Variables<br>
<br>
https://ift.tt/NZPtADS<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<section>
<p><em>Please Leave Feedback in: <a href="https://github.com/rpeszek/rpeszek.github.io/discussions/1" target="_blank">git discussions</a></em></p>
<p>Previous post: <a href="https://rpeszek.github.io/posts/2022-01-03-ts-types-part3.html" target="_blank">Part 3. TS Complexity</a>.</p>
<p><strong>Disclaimers:</strong> (imagine this is a very small font, read it very fast in a half whisper)<br>
<em>I assume strict compiler flags are on, something you get by default with scaffolding, e.g.&nbsp;using <code>create-react-app my-project --template typescript</code> is close enough.<br>
The code examples have been tested with TypeScript v4.5.2.<br>
This post is a pandoc output of a markdown document and code examples are not interactive.<br>
Most of the code examples are published in <a href="https://github.com/rpeszek/ts-experiments/tree/master/ts-notes" target="_blank">ts-notes</a> folder in this github repo: <a href="https://github.com/rpeszek/ts-experiments" target="_blank">ts-experiments</a>.</em></p>
<p><strong>Motivating Quote for the series:</strong></p>
<blockquote>
<p>&ldquo;TypeScript began its life as an attempt to bring traditional object-oriented types to JavaScript so that the programmers at Microsoft could bring traditional object-oriented programs to the web. As it has developed, TypeScript&rsquo;s type system has evolved to model code written by native JavaScripters. The resulting system is <em>powerful, interesting and messy.</em>&rdquo;</p>
</blockquote>
<p><em>From typescriptlang <a href="https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html" target="_blank">TypeScript for Functional Programmers</a></em></p>
<h2>Nutshell</h2>
<p>This is the fourth post in the series devoted to types in TypeScript. In this series, I explore type-centric approaches to writing code and push TS to its limits in doing so. I am writing these posts for like minded developers who are interested in types and either use or consider using TypeScript.</p>
<p>This post will be a little more advanced and will focus on programming with type variables.</p>
<p><a href="https://www.goodreads.com/book/show/112252.Types_and_Programming_Languages" target="_blank">Types and Programming Languages</a> is the book about types I recommend to everyone (&hellip; even if not very successfully). Reading TAPL will be a big eye opener for many developers. The good news is that types dramatically increase programming efficiency so learning them is a good investment.<br>
This section of the post will be a little more TAPL-ish with some more advanced CS. The topics I am about to present are IMO very useful and I will try my best to present them in a digestible way.</p>
<p>I will discuss type variable scoping, rank-2 types, and existential types. Some examples show a level of safety that I did not expect to be able to pull off! As it turns out, we can even prevent subtyping in TS.</p>
<p>Before we start I need to build up some tooling. I will start with a tiny bit of type level programming.</p>
<h2>Safety preventing <code>unknown</code>
</h2>
<p>In previous posts, we have seen examples where TS decided to widen types to <code>unknown</code> rather than report a compilation error.<br>
Interestingly, TS allows enough type level programming so we can try to fix such issues ourselves.</p>
<div>
<pre><code><span>type IsUnknown<span>&lt;</span>T<span>&gt;</span> <span>=</span> unknown <span>extends</span> T<span>?</span> <span>true</span><span>:</span> <span>false</span></span>

<span><span>function</span> verifyUnknown<span>&lt;</span>T<span>&gt;</span>(p<span>:</span> IsUnknown<span>&lt;</span>T<span>&gt;,</span> t<span>:</span> T)<span>:</span> T {</span>
<span> <span>return</span> t</span>
<span>}</span>

<span><span>verifyUnknown</span>(<span>false</span><span>,</span> <span>"test"</span>)</span>
<span><span>const</span> unk<span>:</span> unknown <span>=</span> {}</span>
<span><span>verifyUnknown</span>(<span>true</span><span>,</span> unk)</span></code></pre>
</div>
<div>
<pre><code><span><span>//Compilation Error</span></span>
<span><span>//Argument of type 'false' is not assignable to parameter of type 'true'.ts(2345)</span></span>
<span><span>verifyUnknown</span>(<span>false</span>, unk)</span></code></pre>
</div>
<p>In my first post, I had an example of incorrect code <a href="https://rpeszek.github.io/posts/2021-12-12-ts-types-part1.html#bumps-on-the-path" target="_blank"><code>body4</code></a> inferred as <code>unknown</code> instead of a <code>string</code>. Wrapping such code in <code>verifyUnknown(false, body4)</code> would have alerted me with a compilation error.<br>
You may point out that a much simpler solution is to just type annotate: <code>const body4: string</code>.<br>
I agree. However, having a more generic solution at our disposal is also useful. We will see shortly why.</p>
<p>Here is a short TAPL-ish explanation of what just happened. TS allows me to use type level ternaries. <code>IsUnknown&lt;T&gt;</code> is a type level function (TAPL&rsquo;sh term for this is <em>Type Family</em>) that maps types <code>T</code> to literal boolean types <code>true</code> or <code>false</code>. These types have only a single (a <em>singleton</em>) value: <code>true: true</code> and <code>false: false</code>. If I write <code>verifyUnknown(false, someExpression)</code>, TS will figure out that it has to use <code>false</code> as the type. <code>false</code> matches the second part of the type level ternary and, thus, implies that the ternary predicate <code>unknown extends T</code> is not true. Hence <code>T</code> is not <code>unknown</code>.</p>
<p>I will use <code>verifyUnknown</code> to do some type level trickery. You may wonder if we can extend this approach to other types, not just to <code>unknown</code>. I will get there in this post as well.</p>
<h2>Type variable scoping</h2>
<p>Type variable scoping has two aspects. Let&rsquo;s start with the most obvious one. The type variable being visible inside of the implementation body. This is simple stuff, I just want to share an obvious gotcha that got me at some point:</p>
<div>
<pre><code><span><span>export</span> <span>const</span> bodyScopeExample1 <span>=</span> <span>&lt;</span>T<span>&gt;</span>(value<span>:</span> T <span>|</span> <span>undefined</span> <span>|</span> <span>null</span>)<span>:</span> <span>void</span> <span>=&gt;</span> {</span>
<span> <span>if</span>(value) {</span>
<span> <span>const</span> <span>t</span><span>:</span> T <span>=</span> value <span>//you can access type variables in function body</span></span>
<span> }</span>
<span>}</span></code></pre>
</div>
<p>This approach to defining type signatures (actually my preferred way to write function type signatures) puts <code>T</code> out of scope:</p>
<div>
<pre><code><span>export <span>const</span> bodyScopeExample2: &lt;T&gt;(_: T | undefined | <span>null</span>) =&gt; <span>void</span> = value =&gt; {</span>
<span> <span>if</span>(value) {</span>
<span> <span>const</span> t: T = value <span>//Cannot find name 'T'.ts(2304)</span></span>
<span> }</span>
<span>}</span></code></pre>
</div>
<p>For the type variables to be visible in the implementation they need to be on the <em>RHS</em> of <code>=</code>.</p>
<p>The other aspect of type variable scoping is much more interesting:</p>
<h2>Higher Rank types</h2>
<p>Consider these 2 function declarations:</p>
<div>
<pre><code><span>declare <span>function</span> fn1<span>&lt;</span>T<span>&gt;</span> (f<span>:</span>(t<span>:</span>T)<span>=&gt;</span> <span>void</span>)<span>:</span> <span>void</span> </span>
<span>declare <span>function</span> <span>fn2</span>(f<span>:</span> <span>&lt;</span>T<span>&gt;</span>(t<span>:</span>T)<span>=&gt;</span> <span>void</span>)<span>:</span> <span>void</span> </span></code></pre>
</div>
<p>In <code>fn2</code> the scope of <code>T</code> is much narrower. In TAPL-ish this would be called a rank-2 type.<br>
So what is the difference? Let&rsquo;s try to use both:</p>
<div>
<pre><code><span><span>const</span> useStr <span>=</span> (s<span>:</span>string)<span>:</span> <span>void</span> <span>=&gt;</span> {}</span>
<span><span>fn1</span>(useStr)</span></code></pre>
</div>
<div>
<pre><code><span><span>//Compilation Error:</span></span>
<span><span>//const useStr: (s: string) =&gt; void</span></span>
<span><span>//Argument of type '(s: string) =&gt; void' is not assignable to parameter of type '&lt;T&gt;(t: T) =&gt; void'.</span></span>
<span><span>// Types of parameters 's' and 't' are incompatible.</span></span>
<span><span>// Type 'T' is not assignable to type 'string'.ts(2345)</span></span>
<span><span>fn2</span>(useStr)</span></code></pre>
</div>
<p>Basically <code>fn2</code> requires the argument to be fully generic and <code>useStr</code> is not.</p>
<p>I can play the same games with generic arguments that return <code>T</code></p>
<div>
<pre><code><span>declare <span>function</span> <span>fn4</span>(f<span>:</span> <span>&lt;</span>T<span>&gt;</span>() <span>=&gt;</span> T)<span>:</span> <span>void</span></span></code></pre>
</div>
<p>but will not do that here as these tend to be less practically useful.</p>
<p>Here is how I think about it:</p>
<p><em>Higher rank means generics are first class</em></p>
<h2>Existential types</h2>
<p>In TAPL-ish this is called <em>existential quantification</em> and it has to do with the ownership of definitions. In OO you would say &ldquo;code to interfaces, not implementation&rdquo;, it is also related to the OO concepts of <em>inversion of control</em> and <em>dependency injection</em>. Here is how the story goes:</p>
<h3>Replacing factory pattern</h3>
<div>
<pre><code><span><span>interface</span> Foo {</span>
<span> <span>foo</span><span>:</span> string</span>
<span>}</span>

<span><span>class</span> MyFoo <span>implements</span> Foo{</span>
<span> <span>foo</span><span>:</span> string</span>
<span> <span>constructor</span>() {</span>
<span> <span>this</span><span>.</span><span>foo</span> <span>=</span> <span>"bar"</span></span>
<span> }</span>
<span>}</span></code></pre>
</div>
<p>We want to be able to hide which implementation of <code>Foo</code> we are passing to a callback.<br>
Our first approach tries to use a vanilla TS generic function with a callback argument.<br>
The input function parameter uses some <code>&lt;T extends Foo&gt;</code> of an unknown exact implementation type:</p>
<div>
<pre><code><span>function factoryWithCallback&lt;T <span>extends</span> Foo&gt; (f:(_:T) =&gt; <span>void</span>): <span>void</span> {</span>
<span> <span>//Argument of type 'MyFoo' is not assignable to parameter of type 'T'.</span></span>
<span> <span>// 'MyFoo' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'Foo'.ts(2345)</span></span>
<span> <span>f</span>(<span>new</span> <span>MyFoo</span>()) <span>//Compilation Error</span></span>
<span>}</span></code></pre>
</div>
<p>It does not work and it should not work! We need to use a rank-2 definition:</p>
<div>
<pre><code><span><span>//Compiles!</span></span>
<span><span>function</span> <span>existencialFactory</span>(f<span>:</span> <span>&lt;</span>T <span>extends</span> Foo<span>&gt;</span>(_<span>:</span>T) <span>=&gt;</span> <span>void</span>)<span>:</span> <span>void</span> {</span>
<span> <span>f</span>(<span>new</span> <span>MyFoo</span>())</span>
<span>}</span></code></pre>
</div>
<p>This simulates what is called an existential type. The function that accepts a callback owns the definition of the exact type that is passed to the callback. The callback itself needs to be generic and accept any possible implementation.<br>
Note the scoping of <code>T</code> inside the type defining the function parameter.</p>
<p>This <em>inverts the control</em> from the implementation of the callback to the caller.</p>
<div>
<p><strong>Note on terminology:</strong> Repeating the above definitions</p>
<div>
<pre><code><span>declare <span>function</span> fn1<span>&lt;</span>T<span>&gt;</span> (f<span>:</span>(t<span>:</span>T)<span>=&gt;</span> <span>void</span>)<span>:</span> <span>void</span> </span></code></pre>
</div>
<p>The <code>fn1</code> needs to be defined for all possible types <code>T</code>. The name for it is <em>universal quantification</em>. Some languages even use the <code>forall</code> keyword to describe it.</p>
<div>
<pre><code><span>declare <span>function</span> <span>fn2</span>(f<span>:</span> <span>&lt;</span>T<span>&gt;</span>(t<span>:</span>T)<span>=&gt;</span> <span>void</span>)<span>:</span> <span>void</span> </span></code></pre>
</div>
<p><code>fn2</code> function parameter <code>f</code> needs to be defined for all possible types <code>T</code>. However, <code>fn2</code> can pick whatever type it wants for <code>T</code> and use <code>f</code> with it.<br>
In other words, there exists some type <code>T</code> that will be used but <code>f</code> has no way of knowing which. The name for it is <em>existential quantification</em>. Some languages even use the <code>exists</code> keyword to describe it.</p>
<p>The general concept of existentials is broader than what I am describing here and what TS supports. However, this by itself is plenty powerful.</p>
</div>
<h3>Preventing information escape</h3>
<p>I am drawing a blank trying to think about an OO analogy for this. It is somewhat related to friend classes in C++, package-private scope in Java &hellip; only not exactly.<br>
This example will accomplish more than the above &lsquo;factory&rsquo; pattern and will not use any interfaces or classes:</p>
<div>
<pre><code><span><span>// Using higher rank to protect data</span></span>
<span><span>// Imaginary world without debuggers, JSON.stringify, etc</span></span>
<span>type Api <span>=</span> {<span>getGoodies</span><span>:</span> string[]}</span>

<span><span>//provides access to API, password needs to be protected</span></span>
<span>declare <span>function</span> login<span>&lt;</span>Password<span>&gt;</span>(p<span>:</span> Password)<span>:</span> Api </span>

<span><span>//provide password to a computation, that computation should be able to use the password but shouldn't return it</span></span>
<span><span>const</span> secretive <span>=</span> <span>&lt;</span>R<span>&gt;</span> (fn<span>:</span> <span>&lt;</span>Password<span>&gt;</span> (p<span>:</span> Password) <span>=&gt;</span> R)<span>:</span> R <span>=&gt;</span> {</span>
<span> <span>const</span> <span>s</span> <span>:</span> any <span>=</span> <span>"topsecret"</span></span>
<span> <span>return</span> <span>fn</span> (s)</span>
<span>}</span></code></pre>
</div>
<p>The example is somewhat contrived with the main goal of illustrating the point.<br>
This code exposes building blocks that work together. To get the access to the <code>Api</code> type, you have to use <code>login</code> and you have to use it inside the provided <code>secretive</code> function. Working with an API like this is like assembling a jigsaw puzzle. Types prevent from jamming a square peg into a round hole.<br>
Note, <code>Password</code> is a type variable and we have used the existential type trick.</p>
<p>This code uses the building blocks:</p>
<div>
<pre><code><span><span>const</span> goodProgram <span>=</span> <span>&lt;</span>Password<span>&gt;</span>(p<span>:</span> Password)<span>:</span> string[] <span>=&gt;</span> {</span>
<span> <span>const</span> api <span>=</span> <span>login</span>(p)</span>
<span> <span>return</span> api<span>.</span><span>getGoodies</span></span>
<span>}</span>

<span><span>const</span> stealPassword <span>=</span> <span>&lt;</span>Password<span>&gt;</span>(p<span>:</span> Password)<span>:</span> Password <span>=&gt;</span> p</span>

<span><span>secretive</span>(goodProgram)</span>
<span><span>secretive</span>(stealPassword)</span></code></pre>
</div>
<p>Unfortunately, <code>secretive(stealPassword)</code> compiles. Somewhat typical of TS, instead of providing robust type safety, the compiler infers <code>unknown</code> and accepts my questionable code. Hovering over <code>secretive</code> shows me this:</p>
<div>
<pre><code><span><span>//const secretive: &lt;string[]&gt;(fn: &lt;Password&gt;(p: Password) =&gt; string[]) =&gt; string[]</span></span>
<span><span>secretive</span>(goodProgram)</span>

<span><span>//const secretive: &lt;unknown&gt;(fn: &lt;Password&gt;(p: Password) =&gt; unknown) =&gt; unknown</span></span>
<span><span>secretive</span>(stealPassword)</span></code></pre>
</div>
<p>That is why I have created the <code>verifyUnknown</code> safety in the previous section:</p>
<div>
<pre><code><span><span>const</span> valid <span>=</span> <span>verifyUnknown</span>(<span>false</span><span>,</span> <span>secretive</span>(goodProgram)) <span>//valid: string[]</span></span></code></pre>
</div>
<div>
<pre><code><span><span>//Argument of type 'false' is not assignable to parameter of type 'true'.ts(2345)</span></span>
<span><span>const</span> invalid = <span>verifyUnknown</span>(<span>false</span>, <span>secretive</span>(stealPassword)) <span>//does not compile!</span></span></code></pre>
</div>
<p>To make it a bit nicer we can package <code>verifyUnknown</code> and <code>secretive</code> into one function:</p>
<div>
<pre><code><span><span>const</span> verySecretive <span>=</span> <span>&lt;</span>R<span>&gt;</span> (_<span>:</span> IsUnknown<span>&lt;</span>R<span>&gt;,</span> fn<span>:</span> <span>&lt;</span>Password<span>&gt;</span> (p<span>:</span> Password) <span>=&gt;</span> R)<span>:</span> R <span>=&gt;</span> {</span>
<span> <span>const</span> <span>s</span> <span>:</span> any <span>=</span> <span>"topsecret"</span></span>
<span> <span>return</span> <span>fn</span> (s)</span>
<span> }</span>

<span><span>const</span> valid <span>=</span> <span>verySecretive</span>(<span>false</span><span>,</span> goodProgram) <span>//valid: string[]</span></span></code></pre>
</div>
<div>
<pre><code><span><span>const</span> invalid = <span>verySecretive</span>(<span>false</span>, stealPassword) </span></code></pre>
</div>
<p>This creates some interesting safety. Obviously you could still do a lot of mischief if you wanted to. There is a need for some &lsquo;gentlemen&rsquo;s agreements&rsquo; to not use casting, <code>JSON.stringify</code>, to not use <code>true</code> in <code>verySecretive</code> etc. However, if you think about creating clear contract APIs, this approach could be very powerful.</p>
<p>Existentials are not exactly equivalent to OO. However, using existential types can often accomplish a lot of the same things and often in a cleaner way. Using existentials and disabling OO features like <code>unknown</code> feels a bit contrived, but IMO is still useful. It would be nice if TS provided a cleaner way to disable the use of <code>unknown</code>.<br>
I do not know how robust this type of coding is. I have not played enough with this approach in TS to give you a list of gotchas. In my very limited experience, this seems similar to the rest of TS, TS stops working if I start pushing harder.</p>
<div>
<p><strong>Existentials and higher rank at large:</strong> These concepts have lead to some amazing programming.</p>
<p>For example, existentials are related to dependent pairs (dependent sums) in depenently typed programming languages. Dependent typing provides some very strong types. One example could be lists with a type checked length. You want to be able to use such lists when processing runtime data that can have arbitrary size. That size &lsquo;exists&rsquo; but cannot be known statically at the compile time. This is in essence an existential construction.</p>
<p>Another amazing example is an old (1993) code called <em>State Threads (ST)</em> (currently part of std base library in Haskell). It allows to use a local mutable state to define computations that have to be <em>referentially transparent</em> (I have discussed referential transparency in <a href="https://rpeszek.github.io/posts/2021-12-24-ts-types-part2.html#referential-transparency" target="_blank">Part 2</a>). This is possible because the access to mutate the state cannot escape outside of these computations. ST API remains unchanged since it was created 30 year ago, you can&rsquo;t improve on perfection!</p>
<p>I see higher rank types, mostly rank-2 being used a lot. Having ability to pass generic (polymorphic) functions around is very useful. In my non-TS projects, the problem of &lsquo;separating interface from implementation&rsquo; is typically solved by defining an EDSL (Embedded Domain Specific Language) and an interpreter. Interpreters are polymorphic (generic in TS lingo). Rank-2 types have to be used to make them first class and pass them around.</p>
</div>
<h2>Safety preventing subtyping</h2>
<p>Many TS users have observed the need for this. The term <em>exact type</em> is floating around, I believe <em>flow</em> introduced this name. I have seen solutions like this one being proposed:</p>
<div>
<pre><code><span><span>function</span> exact<span>&lt;</span>T<span>&gt;</span>(item<span>:</span>T)<span>:</span> T {</span>
<span> <span>return</span> item</span>
<span>}</span>

<span>type Hello <span>=</span> {<span>hello</span><span>:</span> string}</span></code></pre>
</div>
<div>
<pre><code><span><span>//Argument of type '{ hello: string; since: number; }' is not assignable to parameter of type 'Hello'.</span></span>
<span><span>// Object literal may only specify known properties, and 'since' does not exist in type 'Hello'.ts(2345)</span></span>
<span>exact&lt;Hello&gt;({hello: <span>"world"</span>, since:<span>2002</span>})</span></code></pre>
</div>
<p>This safety is fragile (and a TS design inconsistency IMO) as the following example shows:</p>
<div>
<pre><code><span><span>const</span> helloSince <span>=</span> {<span>hello</span><span>:</span> <span>"world"</span><span>,</span> <span>since</span><span>:</span><span>2002</span>}</span>

<span>exact<span>&lt;</span>Hello<span>&gt;</span>(helloSince) <span>//complies</span></span></code></pre>
</div>
<p>To create something more robust, here is a code that combines the above <code>unknown</code> verification idea with existentials:</p>
<div>
<pre><code><span>type Same<span>&lt;</span>P<span>,</span>T<span>&gt;</span> <span>=</span> P <span>extends</span> T<span>?</span> (T <span>extends</span> P<span>?</span> <span>true</span><span>:</span> <span>false</span>)<span>:</span> <span>false</span></span>

<span><span>const</span> verifySame <span>=</span> <span>&lt;</span>P<span>&gt;</span> () <span>=&gt;</span> <span>&lt;</span>T<span>&gt;</span> (_<span>:</span> Same<span>&lt;</span>P<span>,</span>T<span>&gt;,</span> t<span>:</span>T)<span>:</span> T <span>=&gt;</span> t</span>

<span>verifySame<span>&lt;</span>Hello<span>&gt;</span>()(<span>true</span><span>,</span> {<span>hello</span><span>:</span> <span>"world"</span>}) <span>//'true' indicates that type matches</span></span>
<span>verifySame<span>&lt;</span>Hello<span>&gt;</span>()(<span>false</span><span>,</span> {<span>hello</span><span>:</span> <span>"world"</span><span>,</span> <span>since</span> <span>:</span> <span>2020</span>}) <span>//'false' is needed to acknowledge types are different </span></span></code></pre>
</div>
<div>
<pre><code><span><span>//Argument of type 'true' is not assignable to parameter of type 'false'.ts(2345)</span></span>
<span>verifySame&lt;Hello&gt;()(<span>true</span>, {hello: <span>"world"</span>, since : <span>2020</span>})</span>
<span>verifySame&lt;Hello&gt;()(<span>true</span>, helloSince)</span></code></pre>
</div>
<p>You may have noticed a case of typing euphoria here. I used rank-2 construction because it allows me to type annotate with only one type variable. This is nice but often not essential.</p>
<p>Here is an implementation of <code>safePush</code> that acts invariant, it does not use any existential tricks:</p>
<div>
<pre><code><span><span>//Note to get 'safePush' I ended up with casting, this is a quick and dirty example and can be done slightly better</span></span>
<span><span>// However, this cast could be an indication that we are changing how TS compiler works</span></span>
<span><span>// Kinda makes sense, to overrule the compiler I may need to cast</span></span>
<span><span>const</span> safePush <span>=</span> <span>&lt;</span>P<span>,</span> T<span>&gt;</span> (_<span>:</span> Same<span>&lt;</span>P<span>,</span>T<span>&gt;,</span> ps<span>:</span> P[]<span>,</span> t<span>:</span> T)<span>:</span> number <span>=&gt;</span> ps<span>.</span><span>push</span>(t <span>as</span> any)</span>

<span><span>const</span> intlist<span>:</span> number[] <span>=</span> [<span>1</span><span>,</span><span>2</span><span>,</span><span>3</span>]</span>
<span><span>const</span> unklist<span>:</span> unknown[] <span>=</span> intlist <span>//exploits array covariance</span></span>
<span>unklist<span>.</span><span>push</span>(<span>"not a number"</span>) <span>//unsafe 'push' adds a 'string' to 'intlist'</span></span>

<span><span>safePush</span>(<span>true</span><span>,</span> intlist<span>,</span> <span>1</span>) <span>//this is safe</span></span></code></pre>
</div>
<div>
<pre><code><span><span>safePush</span>(<span>true</span>, unklist, <span>1</span>) <span>//this is risky and will not compile </span></span>
<span><span>safePush</span>(<span>true</span>, unklist, <span>"not a number"</span>) <span>//this is risky (here wrong) and will not compile </span></span></code></pre>
</div>
<p>Note, to be even safer I would need to prevent <code>unknown</code> as well:</p>
<div>
<pre><code><span><span>const</span> unkstr<span>:</span> unknown <span>=</span> <span>"not a number"</span></span>
<span><span>safePush</span>(<span>true</span><span>,</span> unklist<span>,</span> unkstr) <span>//unfortunately compiles</span></span>

<span><span>//An even safer version of 'Same'</span></span>
<span>type SameAndKnown<span>&lt;</span>P<span>,</span>T<span>&gt;</span> <span>=</span> P <span>extends</span> T<span>?</span> (T <span>extends</span> P<span>?</span> (unknown <span>extends</span> T<span>?</span> <span>false</span><span>:</span> <span>true</span>)<span>:</span> <span>false</span>)<span>:</span> <span>false</span></span>

<span><span>const</span> verySafePush <span>=</span> <span>&lt;</span>P<span>,</span> T<span>&gt;</span> (_<span>:</span> SameAndKnown<span>&lt;</span>P<span>,</span>T<span>&gt;,</span> ps<span>:</span> P[]<span>,</span> t<span>:</span> T)<span>:</span> number <span>=&gt;</span> ps<span>.</span><span>push</span>(t <span>as</span> any)</span>

<span><span>verySafePush</span>(<span>true</span><span>,</span> intlist<span>,</span> <span>1</span>) <span>//this is safe</span></span></code></pre>
</div>
<div>
<pre><code><span><span>verySafePush</span>(<span>true</span>, unklist, unkstr) <span>//this is risky and will not compile!</span></span></code></pre>
</div>
<p>We have discussed problems with the TS approach to variance in the <a href="https://rpeszek.github.io/posts/2022-01-03-ts-types-part3.html#variance-problems" target="_blank">previous installment</a>. We have a DIY approach to fight back!</p>
<p>Side Note: The linked github repo has an existentially typed version of <code>safePush</code> (<code>safePush2</code>) that has just one top level type variable. That version is more cumbersome to use. TS ends up not working well with it.</p>
<p>Another fun exercise:</p>
<div>
<pre><code><span><span>const</span> safeEq <span>=</span> <span>&lt;</span>P<span>,</span> T<span>&gt;</span> (_<span>:</span> Same<span>&lt;</span>P<span>,</span>T<span>&gt;,</span> a<span>:</span> P<span>,</span> b<span>:</span> T)<span>:</span> boolean <span>=&gt;</span> a <span>===</span> (b <span>as</span> unknown)</span>

<span><span>safeEq</span>(<span>true</span><span>,</span> {<span>hello</span><span>:</span> <span>"word"</span>}<span>,</span> {<span>hello</span><span>:</span><span>"dolly"</span>})</span></code></pre>
</div>
<div>
<pre><code><span><span>safeEq</span>(<span>true</span>, {hello: <span>"word"</span>}, {hello:<span>"word"</span>, since:<span>2022</span>}))</span>
<span><span>safeEq</span>(<span>true</span>, <span>1</span>, <span>"str"</span>)</span></code></pre>
</div>
<p>We have discussed problems with TS approach to <code>===</code> narrowing in the <a href="https://rpeszek.github.io/posts/2022-01-03-ts-types-part3.html#complexity-of-ts-types" target="_blank">previous installment</a>. Again, we have a DIY approach to fight back.</p>
<p>This section is related to a number of feature requests: <a href="https://github.com/microsoft/TypeScript/issues/12936" target="_blank">TypeScript issue 12936</a> and <a href="https://github.com/microsoft/TypeScript/issues/7481" target="_blank">TypeScript issue 7481</a>. Hopefully a future version of TS will provide a simpler way to achieve invariance and disable subtyping.</p>
<h2>Phantom types</h2>
<p>TypeScript is somewhat unique in supporting <em>Structural Types</em>. Types like <code>type Person = {firstNm: string, lastNm: string}</code> are structural. That means the name <code>Person</code> is only an alias, what defines the type is the RHS of the definition, not the LHS. Contrast this with an OO class definition in a language like Java. Two structurally identical classes are still considered different types (this is called <em>nominal typing</em>).</p>
<p>It is sometimes convenient to be able to define different types that share the same structure. <em>Phantom types</em> are a way to do that. We say <em>phantom</em> because these types have no impact on runtime values.</p>
<p>Somewhere around 2006, haskell wiki published a write-up about <a href="https://wiki.haskell.org/Phantom_type" target="_blank">phantom types</a>. The write-up was expanded in 2010 to include a form validation example. Since then all blogs (in any programming language) about phantoms show a validation example. I decided to be as unoriginal as everyone else. This will allow me to better focus on how it is done in TS.</p>
<p>My first attempt at phantom types in TS will fail. But this code should make the idea behind phantoms clear:</p>
<div>
<pre><code><span><span>//Marker type</span></span>
<span>type Validated <span>=</span> {<span>type</span><span>:</span> <span>"validated"</span>}</span>

<span><span>//For simplicity this is just a string</span></span>
<span>type ValidationError <span>=</span> string</span>

<span><span>//Extra phantom type variable 'T' </span></span>
<span>type Person<span>&lt;</span>T<span>&gt;</span> <span>=</span> {<span>firstNm</span><span>:</span> string<span>,</span> <span>lastNm</span><span>:</span> string}</span>

<span><span>//Validate person in some way returning 'Validated' phantom marker</span></span>
<span>declare <span>function</span> validate<span>&lt;</span>T<span>&gt;</span>(p<span>:</span> Person<span>&lt;</span>T<span>&gt;</span>)<span>:</span> ValidationError <span>|</span> Person<span>&lt;</span>Validated<span>&gt;</span> </span>

<span><span>//Function to be used only if phantom 'T' is the 'Validated' type </span></span>
<span>declare <span>function</span> <span>doSomethingValidated</span>(p<span>:</span> Person<span>&lt;</span>Validated<span>&gt;</span>)<span>:</span> <span>void</span></span></code></pre>
</div>
<p>Again, these types are trying to create a jigsaw puzzle. One I can assemble in a specific way only.<br>
If the puzzle machinery works, I will have to call <code>validate</code> first to be able to use <code>doSomethingValidated</code>.</p>
<p>Only, this machinery does not work. The following code compiles:</p>
<div>
<pre><code><span><span>function</span> validatedOrNot<span>&lt;</span>T<span>&gt;</span>(p<span>:</span> Person<span>&lt;</span>T<span>&gt;</span>)<span>:</span> <span>void</span> {</span>
<span> <span>doSomethingValidated</span>(p)</span>
<span>}</span>

<span>type ClearlyNotValidated <span>=</span> {<span>type</span><span>:</span> <span>"notvalidated"</span>}</span>

<span><span>function</span> <span>notValidated</span> (p<span>:</span> Person<span>&lt;</span>ClearlyNotValidated<span>&gt;</span>)<span>:</span> <span>void</span> {</span>
<span> <span>doSomethingValidated</span>(p)</span>
<span>}</span></code></pre>
</div>
<p>The fix is to provide a value level information about <code>T</code> in an optional property.<br>
This type definition replaces the one above:</p>
<div>
<pre><code><span><span>//Modified definition adds value level representation `phantom?: T` </span></span>
<span>type Person<span>&lt;</span>T<span>&gt;</span> <span>=</span> {<span>firstNm</span><span>:</span> string<span>,</span> <span>lastNm</span><span>:</span> string<span>,</span> phantom<span>?:</span> T} </span>

<span><span>//provide a way to create person that ignores the additional 'phantom' property:</span></span>
<span><span>const</span> createPerson <span>:</span> <span>&lt;</span>T<span>&gt;</span>(fst<span>:</span> string<span>,</span> lst<span>:</span> string) <span>=&gt;</span> Person<span>&lt;</span>T<span>&gt;</span> <span>=</span> (fst<span>,</span> lst) <span>=&gt;</span> {</span>
<span> <span>return</span> {<span>firstNm</span><span>:</span> fst<span>,</span> <span>lastNm</span><span>:</span> lst}</span>
<span>} </span></code></pre>
</div>
<p>Now this compiles:</p>
<div>
<pre><code><span><span>function</span> <span>validated</span>(p<span>:</span> Person<span>&lt;</span>Validated<span>&gt;</span>)<span>:</span> <span>void</span> {</span>
<span> <span>doSomethingValidated</span>(p)</span>
<span>}</span></code></pre>
</div>
<p>But these no longer do:</p>
<div>
<pre><code><span><span>// Compilation Error</span></span>
<span><span>// Argument of type 'Person&lt;T&gt;' is not assignable to parameter of type 'Person&lt;Validated&gt;'.</span></span>
<span><span>// Type 'T' is not assignable to type 'Validated'.ts(2345)</span></span>
<span>function validatedOrNot&lt;T&gt;(p: Person&lt;T&gt;): <span>void</span>{</span>
<span> <span>doSomethingValidated</span>(p)</span>
<span>}</span>

<span><span>// Compilation Error</span></span>
<span><span>// Argument of type 'Person&lt;ClearlyNotValidated&gt;' is not assignable to parameter of type 'Person&lt;Validated&gt;'.</span></span>
<span><span>// Type 'ClearlyNotValidated' is not assignable to type 'Validated'.</span></span>
<span><span>// Types of property 'type' are incompatible.</span></span>
<span><span>// Type '"notvalidated"' is not assignable to type '"validated"'.ts(2345)</span></span>
<span>function <span>notValidated</span> (p: Person&lt;ClearlyNotValidated&gt;): <span>void</span> {</span>
<span> <span>doSomethingValidated</span>(p)</span>
<span>}</span></code></pre>
</div>
<p>I believe phantom types are used by some FP libraries in TS, e.g.&nbsp;<em>fp-ts</em>, these libraries use somewhat different techniques to get phantoms. There may be advantages to doing phantom types differently than what I have presented. The above approach is the simplest I can think of.</p>
<div>
<p><strong>Phantom types at large:</strong> Phantom types can be used to do a lot of crazy type level stuff. The most wild use I have seen is <a href="https://iohk.io/en/research/library/papers/ghosts-of-departed-proofsfunctional-pearls/" target="_blank">Ghosts of Departed Proofs</a> (this uses Haskell).</p>
<p>Here is a simplified and easy to understand example in TS. Think about a non-mutable list, your function accepts a list and does something with it, your code needs the list to be sorted to work. You can encapsulate this and conservatively sort it just in case (this approach seems not performance optimal), you can document your function by saying that it is the caller responsibility to sort (do developers read documentation?), &hellip; or you can introduce a phantom type:</p>
<div>
<pre><code><span><span>//Sort status as a phantom type, 'List' has type level information about its sort status.</span></span>
<span>type List<span>&lt;</span>T<span>,</span> SortStatus<span>&gt;</span> <span>=</span> <span>...</span></span>

<span><span>interface</span> Comparator<span>&lt;</span>T<span>&gt;</span> {</span>
<span> <span>compare</span> (<span>o1</span><span>:</span> T<span>,</span> <span>o2</span><span>:</span> T)<span>:</span> number</span>
<span>}</span>

<span>declare <span>function</span> sortAscending <span>&lt;</span>T <span>extends</span> Comparator<span>&lt;</span>T<span>&gt;,</span> AnyStatus<span>&gt;</span> (list<span>:</span> List<span>&lt;</span>T<span>,</span> AnyStatus<span>&gt;</span>)<span>:</span> List<span>&lt;</span>T<span>,</span> <span>"ascending"</span><span>&gt;</span></span>

<span>declare <span>function</span> doSomethingWithSortedList <span>&lt;</span>T <span>extends</span> Comparator<span>&lt;</span>T<span>&gt;&gt;</span> (list<span>:</span> List<span>&lt;</span>T<span>,</span> <span>"ascending"</span><span>&gt;</span>)<span>:</span> <span>void</span></span></code></pre>
</div>
<p>Again, notice the types form pieces of a puzzle and can be fitted only in a specific way.<br>
You can think about a <code>sort</code> as something that not only does what it says, but also provides a <em>token</em> to use later to prove that the sort was done. This <em>token</em> is a phantom type. You can think about creating a library that helps orchestrate a similar approach to programming and this is what the linked article talks about.<br>
Many FP programming languages support GADTs, these are very powerful types and limit the popularity of (subsume) phantom typing.<br></p>
</div>
<p>Phantom types could be a very powerful API building tool.<br>
I am sure you can think about many other interesting use cases, &hellip; like state machines.</p>
<h2>Next Chapter</h2>
<p>I want to talk about recursive types and type level programming. It will be more of a review of TS capabilities in these areas.</p>
<p>I need to take a break from writing posts. The next installment will take me longer, maybe a month or a little more, to finish.<br>
Thank you for reading. Happy New Year!</p>
<p>Here is the link: <a href="https://rpeszek.github.io/posts/2022-02-13-ts-types-part5.html" target="_blank">Part 5</a>.</p>
</section>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live.<br>
<br>
September 1, 2022 at 09:21PM
